### PR TITLE
Fancy pull request loading animations

### LINKED ISSUE #2868: Pull requests list loading animations
From https://github.com/desktop/desktop/issues/2642#issuecomment-329179159:

>While the list is being loaded these would have a subtle wave/pulse animation and once the list comes in we'd cross-fade opacity between the skeleton list and the real one and enable the filter list.

### PR DESCRIPTION
**Pull request by joshaber**
Fixes #2868

liceap2

s/o to @niik for reminding me that `key` is a thing that one needs for animations, and for writing the pulse CSS

### CODE REVIEW

**Comment by joshaber** [comment]
@niik I'm gonna assign this to you in case you wanna tweak the timings or easing.

**Comment by niik** [comment]
@joshaber This looks great! I added a few tweaks in #3267 that I'd love to have included here as well

**Comment by joshaber** [comment]
@niik and I both have code in this PR now, so someone else should take a lookyloo 🙏

**Comment by joshaber** [comment]
Friendly bump ☝️ 

giphy-2

File: app/src/ui/lib/list/list.tsx [hunk 1]
```diff
@@ -325,26 +325,24 @@ export class List extends React.Component<IListProps, IListState> {
...
-        newRow = this.props.rowCount - 1
-      }
-    } else {
-      newRow = row + 1
-      if (newRow > this.props.rowCount - 1) {
-        newRow = 0
+  public nextSelectableRow(
+    direction: 'up' | 'down',
+    row: number
+  ): number | null {
```
Comment (shiftkey) [inline app/src/ui/lib/list/list.tsx]: I think this closes out part of https://github.com/desktop/desktop/issues/996

**Comment by donokuda** [comment]
CSS looks legit 2 me but, like @shiftkey, I'm not quite sure how to test this. (I even tried opening the menu with my WiFi turned off 🙈)

**Comment by j-f1** [comment]
> I'm not quite sure how to test this

[This Stack Overflow answer](https://stackoverflow.com/a/32137673/5244995) shows how to simulate bad Internet connections with the Chrome DevTools. Does that help?

**Comment by shiftkey** [comment]
> Does that help?

I've got a sassy comeback here about Australian internets not requiring a simulation, but I'll try out the Chrome Dev Tools too 😉

**Comment by donokuda** [comment]
> This Stack Overflow answer shows how to simulate bad Internet connections with the Chrome DevTools. Does that help?

Ayyup! That does help. It looks like it still shows the empty state before loading the list, just like what @shiftkey stated earlier.

**Comment by shiftkey** [comment]
@donokuda @j-f1 yeah, I'm not sure what's up - what I see doesn't match up with the initial GIF that @joshaber posted. I'm gonna hold up on merging this, just in case we've missed something.

**Comment by joshaber** [comment]
Great catch @shiftkey and @donokuda. Fixing that also surfaced some problems with the pull request store, so this PR got a bit more involved, but we're better for it.

File: app/src/lib/stores/pull-request-store.ts [hunk 2]
```diff
@@ -48,25 +50,48 @@ export class PullRequestStore {
...
-
-      this.pullRequests = await this.getPullRequests(repository)
-      this.emitUpdate()
+      const prs = await this.getPullRequests(repository)
+      await this.refreshStatusForPRs(prs, repository, account)
     } catch (error) {
       log.warn(`Error refreshing pull requests for '${repository.name}'`, error)
       this.emitError(error)
+    } finally {
+      this.changeActiveFetchCount(repository, c => c - 1)
```
Comment (iAmWillShepherd) [inline app/src/lib/stores/pull-request-store.ts]: Shouldn't we only do this if no error is caught?

File: app/src/lib/stores/pull-request-store.ts [hunk 3]
```diff
@@ -48,25 +50,48 @@ export class PullRequestStore {
...
+      this.emitUpdate(repository)
```
Comment (iAmWillShepherd) [inline app/src/lib/stores/pull-request-store.ts]: Is there ever a time we don't want to emit after calling `changeActiveFetchCount`? If not, I think we should move `this.emitUpdate` into `changeActiveFetchCount`.

File: app/src/lib/stores/pull-request-store.ts [hunk 4]
```diff
@@ -39,6 +38,9 @@ export class PullRequestStore {
   ): Promise<void> {
     const api = API.fromAccount(account)
 
+    this.changeActiveFetchCount(repository, c => c + 1)
```
Comment (iAmWillShepherd) [inline app/src/lib/stores/pull-request-store.ts]: Shouldn't this be in the try block given this function can throw an error?

File: app/src/lib/stores/pull-request-store.ts [hunk 5]
```diff
@@ -48,25 +50,48 @@ export class PullRequestStore {
...
+      repository.dbID
+    )
+    const currentCount = this.activeFetchCountPerRepository.get(key) || 0
+    const newCount = fn(currentCount)
+    this.activeFetchCountPerRepository.set(key, newCount)
+  }
+
+  /** Is the store currently fetching the list of open pull requests? */
+  public isFetchingPullRequests(repository: GitHubRepository): boolean {
+    const key = repository.dbID
```
Comment (iAmWillShepherd) [inline app/src/lib/stores/pull-request-store.ts]: We are using `forceUnwrap` above, is there a reason not to use it here?

File: app/src/lib/stores/pull-request-store.ts (hunk 3 above)
Comment (joshaber) [reply app/src/lib/stores/pull-request-store.ts]: Yeah I went back and forth on that. I'm into putting the `emitUpdate` in there 👍

File: app/src/lib/stores/pull-request-store.ts (hunk 4 above)
Comment (joshaber) [reply app/src/lib/stores/pull-request-store.ts]: Are you talking about the `forceUnwrap` call? Exceptions from that function _should_ be fatal as they indicate we failed some invariant.

Comment (iAmWillShepherd) [reply app/src/lib/stores/pull-request-store.ts]: Yes, I was. And thanks for the clarification.

File: app/src/lib/stores/pull-request-store.ts (hunk 5 above)
Comment (joshaber) [reply app/src/lib/stores/pull-request-store.ts]: I was less certain that we could make that guarantee for this call... but thinking about it more, I _think_ we can 👍

**Comment by joshaber** [comment]
🏐

**Comment by iAmWillShepherd** [comment]
I'll let @shiftkey close this out just in case he has more to add.

**Comment by shiftkey** [comment]
Props to @iAmWillShepherd for helping out with the review :gem:

**Pull request merged**